File: rotatedArrayBS.java
class Solution {
    public int search(int[] nums, int target) {
        
        return binarySearch(nums,0 , nums.length-1, target);
    }
    
    public static int binarySearch(int[] nums,int s , int e,  int target){
        
        if(s>e){
            return -1;
        }
        int mid = (s + e)/2;
        if(nums[mid]==target){
            return mid;
        }        
        if(nums[s]<= nums[mid]){
             if (nums[s] <= target && target <= nums[mid]){
                 return binarySearch(nums, s, mid - 1, target);
             }
            else{
                 return binarySearch(nums, mid + 1, e, target);
            }          
        }

        if (nums[mid] <= target && target <= nums[e]){
                 return binarySearch(nums, mid + 1, e, target);
             }
          
        
         return binarySearch(nums, s,mid-1, target);
        
        
        
    }
}

File: rotated-array-BS.py
# // Time Complexity : O(logn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : no 
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        mini, min_index = min(nums), nums.index(min(nums))
        nums_dupe = []
        nums_dupe.extend(nums[min_index:])
        nums_dupe.extend(nums[0:min_index])
        # print(nums_dupe)
        f = 0
        l = len(nums)-1
        while f<=l:
            mid = f + (l - f)//2 # (f+l)//2
            # print(mid, f, l)
            if nums_dupe[mid]>target:
                l = mid - 1
            elif nums_dupe[mid]<target:
                f = mid + 1
            if nums_dupe[mid]==target:
                # print("pk", nums_dupe[mid])
                return (mid+min_index)%len(nums)
        return -1